query-engine/core/src/query_graph_builder/write/delete.rs

```diff
@@ -20,37 +21,86 @@ pub(crate) fn delete_record(
     let where_arg = field.arguments.lookup(args::WHERE).unwrap();
     let filter = extract_unique_filter(where_arg.value.try_into()?, &model)?;
 
-    // Prefetch read query for the delete
-    let mut read_query = read::find_unique(field, model.clone(), query_schema)?;
-    read_query.add_filter(filter.clone());
-
-    let read_node = graph.create_node(Query::Read(read_query));
-    let delete_query = Query::Write(WriteQuery::DeleteRecord(DeleteRecord {
-        model: model.clone(),
-        record_filter: Some(filter.into()),
-    }));
-
-    let delete_node = graph.create_node(delete_query);
-    utils::insert_emulated_on_delete(graph, query_schema, &model, &read_node, &delete_node)?;
-
-    graph.create_edge(
-        &read_node,
-        &delete_node,
-        QueryGraphDependency::ProjectedDataDependency(
-            model.primary_identifier(),
-            Box::new(|delete_node, parent_ids| {
-                if !parent_ids.is_empty() {
-                    Ok(delete_node)
-                } else {
-                    Err(QueryGraphBuilderError::RecordNotFound(
-                        "Record to delete does not exist.".to_owned(),
-                    ))
-                }
+    if can_use_atomic_delete(query_schema, &model, &field) {
+        // Database supports returning the deleted row, so just the delete node will suffice.
+        let nested_fields = field.nested_fields.unwrap().fields;
+        let mut selected_fields = read::utils::collect_selected_scalars(&nested_fields, &model);
+        let selection_order = read::utils::collect_selection_order(&nested_fields);
+
+        let internal_model = &model.dm;
+        let relation_fields = internal_model.fields_pointing_to_model(&model);
+        for relation_field in relation_fields {
+            let parent_relation_field = relation_field.related_field();
+            let linking_fields = parent_relation_field.linking_fields();
+            println!(
+                "laplab: looking at relation field {:?},\ncorresponding parent field is {:?},\nfound linking fields {:?},\nselection before update: {:?}",
+                relation_field, parent_relation_field, linking_fields, selected_fields,
+            );
+            selected_fields = selected_fields.merge(linking_fields);
+            println!("laplab: selection after update: {:?}", selected_fields);
+        }
```

Comment:
Why is this needed? (not a trap question, I'm genuinely wondering)

Reply:
I have added the following comment to this block:
```rust
        // Make sure to request all foreign key fields in addition to the fields selected by the user.
        // This ensures that delete emulation below has all the data it needs, so that the "reload" node
        // is not inserted between delete node and emulation subtree.
        // NOTE: "Reload" node will always throw an error here because we just deleted the record
        // it tries to fetch.
```
Let me know if that clears it up.